helm: Fix duplicate tunnel-protocol configuration in cilium-config configmap

jonasbadstuebner opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

---

We had trouble setting this up too, and as @haribhusal2025 stated in https://github.com/cilium/cilium/issues/33456#issuecomment-2233697293, setting the routingMode and tunnelProtocol explicitly works around the helm templating issues.

I am aware of #34315 attempting to fix this issue, but I think the approach of using a variable and overwriting it, is not the best for user experience.

As a heavy user of helm, having to look at the helm template itself to know what is happening or where values come from is more frustrating than simply being shown an error.
So I propose a solution to the problem that gives an early error message, without introducing more edge cases to check for with nested if-statements.

Fixes: #33456 
Closes: #34315

```release-note
Adjust verification for tunnel-protocol and routing-mode in helm templates to remove occurrence of duplicate entries in rendered configmap.
Remove constraint on tunnelProtocol for aksbyocni.
```

jonasbadstuebner committed:
helm: Fix duplicate tunnel-protocol key in cilium-config configmap

Previously it was only possible to enable geneve with tunnel routing
mode by explicitly setting both as helm values.
Otherwise it would fail like this:

line 132: mapping key "tunnel-protocol" already defined at line 131

Fixes #33456

Signed-off-by: Jonas Badstübner <jonas@jb.software>

gandro commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -485,23 +485,21 @@ data:
+  {{- if ne .Values.routingMode "native" }}
     {{- fail (printf "RoutingMode must be set to native when gke.enabled=true" )}}
   {{- end }}
-  routing-mode: "native"
```
So this is a breaking change in behavior. Previously users were able to install GKE with just `set gke.enabled=true`, but now they explicitly also have to set `--set routingMode=native`. There is already [docs](https://docs.cilium.io/en/stable/installation/k8s-install-helm/) and [tooling](https://github.com/cilium/cilium/blob/fbd276676e7659b986e468d1e428065ff2f9af2a/cilium-cli/install/helm.go#L75-L77) out there that doesn't do that, all of that will be broken and would have to be fixed.

Is there no other way were we can preserve the current behavior without breaking this?

jonasbadstuebner replied:
I could revert this line and add an `and` statement when checking for the routing-mode being set. We are lucky here that gke.enabled is the only exception that requires a hard-coded routing-mode.

gandro replied:
Sounds good to me!

jonasbadstuebner replied:
What do you think of my new proposal?

jonasbadstuebner replied:
enabling `aksbyocni` just enforces the default values. And you cannot overwrite the routing-mode with gke enabled. If you set it to `native` explicitly, it will just be set on line 491 (that's why I have to disable it on line 502, to not let it be set twice), other values are not allowed at all.

gandro reviewed (changes requested):
Thanks for the PR! Unfortunately this is a breaking change - we either need to come up with a solution that doesn't break existing installation commands - or patch all the docs and tooling before we can ship this PR.

gandro commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -494,20 +494,17 @@ data:
-{{- if .Values.tunnelProtocol }}
-  tunnel-protocol: {{ .Values.tunnelProtocol | quote }}
+{{- if not .Values.gke.enabled }}
+  routing-mode: {{ .Values.routingMode | default "tunnel" | quote }}
```
nit: Would it make sense to define a `$fallbackRoutingMode` variable on the top of the file that defaults to "tunnel" but is set to "native" in the GKE case? I think that would be a bit easier to read

jonasbadstuebner replied:
In the configuration I proposed, there are only 2 things to check:
(1) Do I have `gke.enabled: true`? If so, the routing-mode is hardcoded to "native".
If not: (2) Do I have `routingMode` set explicitly? If so, routing-mode is that value, otherwise it is "tunnel".

I think introducing a variable would increase the mental overhead required to understand where the value actually comes from and does not provide much benefit when we only have that one "edge case"-y if-statement for GKE in which the `routing-mode` is specified directly.

jonasbadstuebner replied:
If you disagree, I can change it to the variable, you are the boss.

giorio94 replied:
Yet another possibility could be:

```
routing-mode: {{ .Values.routingMode | default (ternary "native" "tunnel" .Values.gke.enabled) | quote }}
```

jonasbadstuebner replied:
That would not use a variable and only have one place where the routing-mode is defined. I like it.

gandro replied:
That also looks good to me. Let's do that.

jonasbadstuebner replied:
Done, please resolve the conversation if it fits

gandro reviewed (commented):
This looks better to me! I think we should also get a review from @giorio94 to make sure we're fine with `tunnel-protocol: vxlan` if the user specifies an explicitly sets the native routing mode (I believe it should be fine since the agent also defaults to that)

giorio94 commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -494,20 +494,17 @@ data:
-  tunnel-protocol: "vxlan"
+  {{- if ne (.Values.tunnelProtocol | default "vxlan" ) "vxlan" }}
+    {{- fail (printf "TunnelProtocol must be set to vxlan when aksbyocni.enabled=true" )}}
+  {{- end }}
```
I wonder if this is an actual constraint, or any tunnel protocol would work correctly.

jonasbadstuebner replied:
https://github.com/cilium/cilium/pull/34315#issuecomment-2284490286 states:
> `aksbyocni` supports all tunneling protocols IIRC

So maybe we can remove that? But I could not take that decision. It is not the direct scope of this PR - but if you want to make the call to remove this constraint, this could be done easily.

giorio94 replied:
Let's wait on Sebastian's thoughts first, but I'd be personally in favor of dropping the check to keep the logic as simple as possible and limit exceptions.

gandro replied:
Yeah let's drop that check and just ensure that we're running in any tunnel mode. I don't see any indication in the code that it only supports VXLAN. 

~It is possible that this check is a leftover from where `routing-mode` and `tunnel-protocol` where one single flag called `tunnel` and what it actually shoudl check is `routing-mode` (i.e. do the opposite of what GKE does, requring `routing-mode: tunnel`)~ Edit I just realized the routing-mode check is above. Yeah let's drop this :+1:

jonasbadstuebner replied:
Dropped and added the reference to the split of settings, as you mentioned it. If this is incorrect or misleading, I will remove this.

gandro replied:
Looks good! Thanks for adding the release note

giorio94 reviewed (commented):
Thanks. Changes look good to me, just a couple of comments inline.

> I think we should also get a review from @giorio94 to make sure we're fine with `tunnel-protocol: vxlan` if the user specifies an explicitly sets the native routing mode (I believe it should be fine since the agent also defaults to that)

Yes, the agent defaults to vxlan as tunnel protocol in all cases, so that's consistent.

https://github.com/cilium/cilium/blob/89c98fb36c50498e2cf698b10b93566a7109ef59/pkg/datapath/tunnel/tunnel.go#L235-L238

https://github.com/cilium/cilium/blob/89c98fb36c50498e2cf698b10b93566a7109ef59/pkg/defaults/defaults.go#L514-L515

gandro reviewed (approved):
Looks good to me now, thank you! I think we can drop the aksbyocni check too

After that, let's see what CI says

jonasbadstuebner committed:
helm: Remove tunnelProtocol limitation to vxlan for aksbyocni

Previously the tunnel-protocol was limited to VXLAN, when enabling
aksbyocni. As discussed in
https://github.com/cilium/cilium/pull/36226#discussion_r1862119360 this
seems to be a leftover from when there was just one flag called "tunnel"
that combined routing-mode and tunnel-protocol.

This patch makes it possible to configure any tunnel-protocol with
aksbyocni.enabled: true.

Related: #24561

Signed-off-by: Jonas Badstübner <jonas@jb.software>
